cmd/kubeadm/app/phases/selfhosting/selfhosting.go

```diff
@@ -0,0 +1,143 @@
+)
+
+// CreateSelfHostedControlPlane is responsible for turning a Static Pod-hosted control plane to a self-hosted one
+// It achieves that task this way:
+// 1. Load the Static Pod specification from disk (from /etc/kubernetes/manifests)
+// 2. Extract the PodSpec from that Static Pod specification
+// 3. Mutate the PodSpec to be compatible with self-hosting (add the right labels, taints, etc. so it can schedule correctly)
+// 4. Build a new DaemonSet object for the self-hosted component in question. Use the PodSpec above mentioned PodSpec
+// 5. Create the DaemonSet resource. Wait for the Pod to become Running.
+// 6. Remove the Static Pod manifest file. The kubelet will stop the original Static Pod-hosted component that was running.
+// 7. The self-hosted container should now step up and take over.
+// 8. In order to avoid race conditions, we're still making sure the API /healthz endpoint is healthy
+// 9. Do that for the kube-apiserver, kube-controller-manager and kube-scheduler in a loop
+func CreateSelfHostedControlPlane(client *clientset.Clientset) error {
+	components := []string{kubeAPIServer, kubeControllerManager, kubeScheduler}
+
+	for _, componentName := range components {
+		start := time.Now()
+		manifestPath := buildStaticManifestFilepath(componentName)
+
+		// Load the Static Pod file in order to be able to create a self-hosted variant of that file
+		podSpec, err := loadPodSpecFromFile(manifestPath)
+		if err != nil {
+			return err
+		}
+
+		// Build a DaemonSet object from the loaded PodSpec
+		ds := buildDaemonSet(componentName, podSpec)
+
+		// Create the DaemonSet in the API Server
+		if _, err := client.ExtensionsV1beta1().DaemonSets(metav1.NamespaceSystem).Create(ds); err != nil {
+			return fmt.Errorf("failed to create self-hosted %q daemonset [%v]", componentName, err)
+		}
+
+		// Wait for the self-hosted component to come up
+		kubeadmutil.WaitForPodsWithLabel(client, componentName, buildSelfHostedWorkloadLabelQuery(componentName))
+
+		// Remove the old Static Pod manifest
+		if err := os.RemoveAll(manifestPath); err != nil {
+			return fmt.Errorf("unable to delete static pod manifest for %s [%v]", componentName, err)
+		}
+
+		// Make sure the API is responsive at /healthz
+		kubeadmutil.WaitForAPI(client)
+
+		fmt.Printf("[self-hosted] self-hosted %s ready after %f seconds\n", componentName, time.Since(start).Seconds())
+	}
+	return nil
+}
+
+// buildDaemonSet is responsible for mutating the PodSpec and return a DaemonSet based suitable for the self-hosting purporse
+func buildDaemonSet(name string, podSpec *v1.PodSpec) *extensions.DaemonSet {
+	// Mutate the PodSpec so it's suitable for self-hosting
+	mutatePodSpec(name, podSpec)
+
+	// Return a DaemonSet based on that Spec
+	return &extensions.DaemonSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "self-hosted-" + name,
+			Namespace: metav1.NamespaceSystem,
+			Labels: map[string]string{
+				"k8s-app": selfHostingLabelValuePrefix + name,
```

Comment:
I don't think so, as I don't recognize any practical purpose for it
The less code the better